diff --git a/include/harness.h b/include/harness.h
new file mode 100644
index 000000000..d87ad9e15
--- /dev/null
+++ b/include/harness.h
@@ -0,0 +1,109 @@
+#ifndef __PROFTPD_FUZZING_HARNESS__
+#define __PROFTPD_FUZZING_HARNESS__
+
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+
+
+// Stub out all network-related functions.
+#define listen(sockfd, backlog) ( \
+    listen(-1, backlog), \
+    errno = 0, \
+    0 \
+)
+
+#define setsockopt(sockfd, level, optname, optval, optlen) ( \
+    setsockopt(-1, level, optname, optval, optlen), \
+    errno = 0, \
+    0 \
+)
+
+#define socket(domain, type, protocol) ( \
+    socket(-1, type, protocol), \
+    errno = 0, \
+    0 \
+)
+
+#define accept(sockfd, addr, addrlen) ( \
+    accept(-1, addr, addrlen), \
+    errno = 0, \
+    common_sock_stub(sockfd, addr, addrlen) \
+)
+
+#define bind(sockfd, addr, addrlen) ( \
+    bind(-1, addr, addrlen), \
+    errno = 0, \
+    common_sock_stub(sockfd, addr, addrlen) \
+)
+
+#define getpeername(sockfd, addr, addrlen) ( \
+    getpeername(-1, addr, addrlen), \
+    errno = 0, \
+    common_sock_stub(sockfd, addr, addrlen) \
+)
+
+#define getsockname(sockfd, addr, addrlen) ( \
+    getsockname(-1, addr, addrlen), \
+    errno = 0, \
+    common_sock_stub(sockfd, addr, addrlen) \
+)
+
+#define common_sock_stub(sockfd, addr, addrlen) \
+    ( \
+        ((struct sockaddr*)addr) != NULL ? \
+        ( \
+            ((struct sockaddr_in*)addr)->sin_family = AF_INET, \
+            ((struct sockaddr_in*)addr)->sin_addr.s_addr = inet_addr("127.0.0.1"), \
+            ((struct sockaddr_in*)addr)->sin_port = htons(8000), \
+            0 \
+        ) : 0, \
+        0 \
+    )
+
+// Stub out various other functions.
+#define alarm(seconds) ( \
+    alarm(0), \
+    0 \
+)
+
+#define close(fd) \
+    ( \
+        close(-1), \
+        errno = 0, \
+        0 \
+    )
+
+#define dup2(old, new) \
+    ((new) <= 2 ? ( \
+        dup2(old, new + 10), \
+        errno = 0, \
+        0 \
+    ) : dup2(old, new))
+
+#define fcntl(fd, op, ...) ( \
+    fcntl(-1, op), \
+    errno = 0, \
+    0 \
+)
+
+#define fstat(fd, statbuf) \
+    (fd) <= 2 ? ( \
+        fstat(-1, statbuf), \
+        errno = 0, \
+        (statbuf)->st_mode |= S_IFSOCK, \
+        0 \
+    ) : fstat(fd, statbuf)
+
+#define fork() 0
+
+#define nanosleep(duration, rem) ( \
+    nanosleep(0, rem), \
+    0 \
+)
+
+#define write(fd, buf, count) \
+    fd >= 0 ? write(1, buf, count) : write(fd, buf, count)
+
+
+#endif  // __PROFTPD_FUZZING_HARNESS__
diff --git a/modules/mod_rlimit.c b/modules/mod_rlimit.c
index 420bc5314..cf7d6d4df 100644
--- a/modules/mod_rlimit.c
+++ b/modules/mod_rlimit.c
@@ -757,10 +757,12 @@ static void rlimit_postparse_ev(const void *event_data, void *user_data) {
    * resource limits; we would prevent future session processes.
    */
 
+  /*
   rlimit_set_core(DAEMON_SCOPE);
   rlimit_set_cpu(DAEMON_SCOPE);
   rlimit_set_memory(DAEMON_SCOPE);
   rlimit_set_files(DAEMON_SCOPE);
+  */
 }
 
 /* Module initialization */
diff --git a/src/fsio.c b/src/fsio.c
index 584598477..7ed484e08 100644
--- a/src/fsio.c
+++ b/src/fsio.c
@@ -30,6 +30,7 @@
 
 #include "conf.h"
 #include "privs.h"
+#include "harness.h"
 
 #ifdef HAVE_SYS_STATVFS_H
 # include <sys/statvfs.h>
diff --git a/src/inet.c b/src/inet.c
index e9abae0ee..ca4ce5193 100644
--- a/src/inet.c
+++ b/src/inet.c
@@ -28,6 +28,7 @@
 
 #include "conf.h"
 #include "privs.h"
+#include "harness.h"
 
 extern unsigned char is_master;
 extern server_rec *main_server;
@@ -420,7 +421,7 @@ static conn_t *init_conn(pool *p, int fd, pr_netaddr_t *bind_addr,
         pr_signals_unblock();
       }
 
-      pr_timer_sleep(1);
+      //pr_timer_sleep(1);
 
       if (port != INPORT_ANY &&
           port < 1024) {
@@ -540,7 +541,7 @@ conn_t *pr_inet_create_conn_portrange(pool *p, pr_netaddr_t *bind_addr,
 
       if (attempt == 3) {
 	/* Obtain a random index into the port array range. */
-	random_index = (int) ((1.0 * i * rand()) / (RAND_MAX+1.0));
+	random_index = 0; //(int) ((1.0 * i * rand()) / (RAND_MAX+1.0));
 
 	/* Copy the port at that index into the array from which port
 	 * numbers will be selected when calling init_conn().
@@ -774,30 +775,6 @@ int pr_inet_set_proto_opts(pool *p, conn_t *c, int mss, int nodelay,
   }
 #endif /* IP_TOS */
 
-#ifdef IPV6_TCLASS
-  if (pr_netaddr_use_ipv6()) {
-    /* Only set TCLASS flags on IPv6 sockets; IPv4 sockets use TOS. */
-    if (pr_netaddr_get_family(c->local_addr) == AF_INET6) {
-      if (c->listen_fd != -1) {
-        int level, res;
-
-        level = ipv6_proto;
-        res = setsockopt(c->listen_fd, level, IPV6_TCLASS, (void *) &tos,
-          sizeof(tos));
-        if (res < 0
-            && errno != EINVAL
-#ifdef ENOPROTOOPT
-            && errno != ENOPROTOOPT
-#endif /* !ENOPROTOOPT */
-          ) {
-          pr_log_pri(PR_LOG_NOTICE, "error setting listen fd IPV6_TCLASS: %s",
-            strerror(errno));
-        }
-      }
-    }
-  }
-#endif /* IPV6_TCLASS */
-
   if (c->listen_fd != -1) {
     if (pr_inet_set_proto_cork(c->listen_fd, nopush) < 0) {
       pr_log_pri(PR_LOG_NOTICE, "error corking listen fd %d: %s", c->listen_fd,
diff --git a/src/main.c b/src/main.c
index 47d7d6991..453f8b301 100644
--- a/src/main.c
+++ b/src/main.c
@@ -43,6 +43,8 @@
 #endif
 
 #include "privs.h"
+#include "/root/rosa/fuzzers/aflpp/rosa-trace.h"
+#include "harness.h"
 
 int (*cmd_auth_chk)(cmd_rec *);
 void (*cmd_handler)(server_rec *, conn_t *);
@@ -774,6 +776,12 @@ static cmd_rec *make_ftp_cmd(pool *p, char *buf, int flags) {
 
 static void cmd_loop(server_rec *server, conn_t *c) {
 
+#ifdef __AFL_HAVE_MANUAL_CONTROL
+  __AFL_INIT();
+#endif
+
+  __ROSA_TRACE_START();
+
   while (TRUE) {
     int res = 0; 
     cmd_rec *cmd = NULL;
diff --git a/src/netio.c b/src/netio.c
index b505ce666..4b6b53e8b 100644
--- a/src/netio.c
+++ b/src/netio.c
@@ -213,7 +213,7 @@ static int core_netio_postopen_cb(pr_netio_stream_t *nstrm) {
 
 static int core_netio_read_cb(pr_netio_stream_t *nstrm, char *buf,
     size_t buflen) {
-  return read(nstrm->strm_fd, buf, buflen);
+  return read(0, buf, buflen);
 }
 
 static pr_netio_stream_t *core_netio_reopen_cb(pr_netio_stream_t *nstrm, int fd,
