diff --git a/plugins/sudoers/check.c b/plugins/sudoers/check.c
index 6f5d9cc0a..936b2234a 100644
--- a/plugins/sudoers/check.c
+++ b/plugins/sudoers/check.c
@@ -173,6 +173,13 @@ check_user(int validated, int mode)
     if (sudo_auth_init(closure.auth_pw, mode) == -1)
 	goto done;
 
+/*
+ * NOTE: for fuzzing, we need to run sudo AS sudo. But at the same time, we
+ * need to hit the password check. In order to do that, we need to disable
+ * this entire piece of code that skips the password check if the user is
+ * already sudo, or if they're not elevating privileges.
+ */
+#if 0
     /*
      * Don't prompt for the root passwd or if the user is exempt.
      * If the user is not changing uid/gid, no need for a password.
@@ -203,6 +210,7 @@ check_user(int validated, int mode)
 	    goto done;
 	}
     }
+#endif
 
     ret = check_user_interactive(validated, mode, &closure);
 
diff --git a/src/sudo.c b/src/sudo.c
index 0121586e3..303d3d120 100644
--- a/src/sudo.c
+++ b/src/sudo.c
@@ -68,6 +68,7 @@
 #include "sudo.h"
 #include "sudo_plugin.h"
 #include "sudo_plugin_int.h"
+#include "/root/rosa/fuzzers/aflpp/rosa-trace.h"
 
 /*
  * Local variables
@@ -149,6 +150,7 @@ main(int argc, char *argv[], char *envp[])
     const char * const allowed_prognames[] = { "sudo", "sudoedit", NULL };
     sigset_t mask;
     debug_decl_vars(main, SUDO_DEBUG_MAIN);
+    __ROSA_TRACE_START();
 
     /* Only allow "sudo" or "sudoedit" as the program name. */
     initprogname2(argc > 0 ? argv[0] : "sudo", allowed_prognames);
