--- a/test/ossfuzz.c
+++ b/test/ossfuzz.c
@@ -8,6 +8,9 @@
 #endif
 #include <stdio.h>
 #include <string.h>
+#include "/root/rosa/fuzzers/aflpp/rosa-trace.h"
+#include <assert.h>
+#include <stdlib.h>
 #include "sqlite3.h"
 
 #if defined(_MSC_VER)
@@ -122,6 +125,7 @@ int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
   int rc;                  /* Return code from various interfaces */
   char *zSql;              /* Zero-terminated copy of data[] */
   FuzzCtx cx;              /* Fuzzing context */
+  __ROSA_TRACE_START();
 
   memset(&cx, 0, sizeof(cx));
   if( size<3 ) return 0;   /* Early out if unsufficient data */
